pkg/networkservice/common/heal/server.go

```diff
@@ -0,0 +1,439 @@
+
+	conn, err := next.Server(ctx).Request(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check heal client is in the chain and connection was added
+	if _, ok := f.clients.Load(conn.GetId()); !ok {
+		log.FromContext(ctx).WithField("healServer", "Request").Errorf("Heal server ignored connection %s: heal client is not active", conn.GetId())
+		return conn, nil
+	}
+
+	ctx = f.applyStoredCandidates(ctx, conn)
+
+	err = f.startHeal(ctx, request.Clone().SetRequestConnection(conn.Clone()))
+	if err != nil {
+		return nil, err
+	}
+
+	f.conns.Store(conn.GetId(), connection{
+		Connection: conn.Clone(),
+		ctx:        ctx,
+	})
+
+	return conn, nil
+}
+
+func (f *healServer) Close(ctx context.Context, conn *networkservice.Connection) (*empty.Empty, error) {
+	// Replace path within connection to the healed one
+	f.replaceConnectionPath(conn)
+
+	f.stopHeal(conn)
+	f.clients.Delete(conn.Id)
+
+	rv, err := next.Server(ctx).Close(ctx, conn)
+	if err != nil {
+		return nil, err
+	}
+	return rv, nil
+}
+
+func (f *healServer) RegisterClient(ctx context.Context, conn *networkservice.Connection, cc networkservice.MonitorConnectionClient) {
+	f.clients.Store(conn.Id, clientConnInfo{
+		ctx: ctx,
+		cc:  cc,
+	})
+}
+
+func (f *healServer) stopHeal(conn *networkservice.Connection) {
+	var cancelHeal func()
+	<-f.cancelHealMapExecutor.AsyncExec(func() {
+		if cancelHealEntry, ok := f.cancelHealMap[conn.GetId()]; ok {
+			cancelHeal = cancelHealEntry.cancel
+			delete(f.cancelHealMap, conn.GetId())
+		}
+	})
+	if cancelHeal != nil {
+		cancelHeal()
+	}
+	f.conns.Delete(conn.GetId())
+}
+
+// startHeal - start a healAsNeeded using the request as the request for re-request if healing is needed.
+func (f *healServer) startHeal(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) error {
+	errCh := make(chan error, 1)
+	go f.healAsNeeded(ctx, request, errCh, opts...)
+	return <-errCh
+}
+
+// healAsNeeded - heal the connection found in request.  Will immediately do a monitor to make sure the server has the
+// expected connection and it is sane, returning an error via errCh if there is an issue, and nil via errCh if there is
+// not.  You will only *ever* receive one real error via the errCh.  errCh will be closed when healAsNeeded is finished
+// allowing it to double as a 'done' channel we can use when we stopHealing in f.Close.
+// healAsNeeded will then continue to monitor the servers opinions about the state of the connection until either
+// expireTime has passed or stopHeal is called (as in Close) or a different pathSegment is found via monitoring
+// indicating that a later Request has occurred and in doing so created its own healAsNeeded and so we can stop this one
+func (f *healServer) healAsNeeded(baseCtx context.Context, request *networkservice.NetworkServiceRequest, errCh chan error, opts ...grpc.CallOption) {
+	// When we are done, close the errCh
+	defer close(errCh)
```

Comment:
Do we actually need this deferred close? We never wait for the `healAsNeeded` to be fully done and all error branches actually contain some `errCh < err` logic.

Reply:
Done